opentelemetry-sdk/src/metrics/internal/sum.rs

```diff
@@ -331,29 +365,30 @@ impl<T: Number<T>> PrecomputedSum<T> {
 
         if self
             .value_map
-            .has_no_value_attribute_value
+            .has_no_attribute_value
             .swap(false, Ordering::AcqRel)
         {
+            let value = self.value_map.no_attribute_value.get_value();
+            let delta = value - *reported.get(&vec![]).unwrap_or(&T::default());
+            new_reported.insert(vec![], value);
+
             s_data.data_points.push(DataPoint {
                 attributes: vec![],
                 start_time: Some(prev_start),
                 time: Some(t),
-                value: self.value_map.no_attribute_value.get_and_reset_value(),
+                value: delta,
                 exemplars: vec![],
             });
         }
 
-        let mut values = match self.value_map.values.write() {
+        let mut values = match self.value_map.trackers.write() {
             Ok(v) => v,
             Err(_) => return (0, None),
         };
 
-        let default = T::default();
         for (attrs, value) in values.drain() {
-            let delta = value.get_value() - *reported.get(&attrs).unwrap_or(&default);
-            if delta != default {
-                new_reported.insert(attrs.clone(), value.get_value());
-            }
+            let delta = value.get_value() - *reported.get(&attrs).unwrap_or(&T::default());
+            new_reported.insert(attrs.clone(), value.get_value());
```

Comment:
Similar to the no-attributes case, we shouldn't do a new atomic read on `value` here.

Reply:
@utpilla I believe this looks fine, as the hashmap is locked (in line 384) here during collect cycle, so there won't be any inserts happening.

Reply:
Ah yes.

Reply:
We could still avoid paying the cost of another atomic read. Since, we take a lock, we are certain that the value of the variable wouldn't change from what we get on the first atomic read (on line 390). We could simply use that value.

Not a major perf benefit though as this is not the hot path. 😄

Reply:
Ah didn't see this one, fixed.